<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:macro name="kuka_rsi_driver_control_joint" params="name">
    <joint name="${name}">
      <command_interface name="position" />
      <state_interface name="position" />
      <state_interface name="effort" />
    </joint>
  </xacro:macro>

  <xacro:macro name="kuka_rsi_driver_control" params="
    name
    prefix
    rsi_listen_ip
    rsi_listen_port:=49152
    ">
    <ros2_control name="${name}" type="system">
      <hardware>
        <plugin>kuka_rsi_driver/KukaRsiHardwareInterface</plugin>
        <param name="listen_address">${rsi_listen_ip}</param>
        <param name="listen_port">${rsi_listen_port}</param>
      </hardware>

      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a1" />
      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a2" />
      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a3" />
      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a4" />
      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a5" />
      <xacro:kuka_rsi_driver_control_joint name="${prefix}joint_a6" />

      <sensor name="${prefix}tcp">
        <state_interface name="position.x"/>
        <state_interface name="position.y"/>
        <state_interface name="position.z"/>
        <state_interface name="orientation.x"/>
        <state_interface name="orientation.y"/>
        <state_interface name="orientation.z"/>
        <state_interface name="orientation.w"/>
      </sensor>

      <gpio name="${prefix}robot_state">
        <state_interface name="program_state" />
      </gpio>

      <gpio name="${prefix}speed_scaling">
        <state_interface name="speed_scaling_factor" />
      </gpio>

    </ros2_control>
  </xacro:macro>

</robot>
